strawberryfields/engine.py

```diff
@@ -277,15 +272,57 @@ def _broadcast_nones(val, dim):
                 p = p.compile(target, **compile_options)
             p.lock()
 
-            self._run_program(p, **kwargs)
-            shots = kwargs.get("shots", 1)
-            self.samples = [_broadcast_nones(p.reg_refs[k].val, shots) for k in sorted(p.reg_refs)]
+            _, values = self._run_program(p, **kwargs)
             self.run_progs.append(p)
 
+            # check number of measurments and combine samples if needed
+            if len(values) > 1:
+                self._combine_samples(values, p)
+            elif len(values) == 1:
+                self.samples = values[0]
+            else:  # if no measurement was made
+                self.samples = np.array([[]])
+
             prev = p
 
-        if self.samples is not None:
-            return Result(np.array(self.samples).T)
+        return Result(self.samples)
+
+    def _combine_samples(self, values, program):
+        """Combines samples if multiple measurements were made.
+
+        If batches are used, the samples will be fed to ``Result.combine_samples``
+        one-batch-at-a-time, which combines the samples and returns them with shape
+        ``(batches, shots, measured_modes)``.
+
+        Args:
+            values (list[array, tensor]): the sample measurements in a list
+            program (Program): quantum program that is run
+        """
+        # get nested order of measured modes
+        sort_order = [
+            np.array([r.ind for r in c.reg]) for c in program.circuit if "Measure" in c.op.__str__()
+        ]
+
+        # check for duplicate mode-measures
+        if len(np.hstack(sort_order)) != len(set(np.hstack(sort_order))):
+            raise RuntimeError("Modes can only be measured once inside a circuit.")
```

Comment:
Is this true? For example, what about the following case?
```python
with prog.context as q:
    Sgate(0.1) | q[0]
    MeasureX | q[0]  # resets q[0] to the vacuum
    Xgate(q[0].par) | q[0]
    MeasureFock() | q[0]
```

Reply:
I added this since I thought it made sense only measuring once per mode per circuit. As you mentioned further down this can, and maybe should, be fixed so that only the last measurement is saved.

Reply:
Personally I find it confusing to allow multiple measurements on the same mode, even if there is a sensible logic behind it that can be explained further. I'm on board with raising an error to inform the user about it.

The way I usually try to picture these circuits for myself is to think of a circuit diagram. In that sense, I agree with how SF Interactive now deals with measurements: once there was a measurement operator placed on a single wire, no other operations can be placed on that wire anymore. After all, the measurement should result in a "collapse of the wavefunction" on the related wire, right?

Therefore having the logic of resetting to vacuum once a measurement has been done simply would imply to me that we are trying to consider two different circuits.

```python
with prog.context as q:
    Sgate(0.1) | q[0]
    MeasureX | q[0]
```
```python
with prog.context as q:
    Xgate(q[0].par) | q[0]
    MeasureFock() | q[0]
```

Reply:
If I remember, one of the reasons for allowing a mode to be reused after measurement is so that the user had some way to avoid adding ancillas (which can be costly in the fock space simulations).

Regardless of whether we think this feature makes sense or not, is it truly in-scope for this particular PR to change it?